radixdlt/src/main/java/com/radixdlt/consensus/safety/SafetyRules.java

```diff
@@ -66,46 +56,38 @@ public SafetyRules(@Named("self") ECKeyPair selfKey,
 	 * @return the just-committed vertex id, if any
 	 */
 	public Optional<Hash> process(QuorumCertificate qc) {
+		final Builder safetyStateBuilder = this.state.toBuilder();
+
 		// pre-commit phase on vertex's parent if there is a newer consecutive 1-chain
 		// keep highest 1-chain as the current "generic" QC
 		if (qc.getView().compareTo(this.state.getGenericView().orElse(View.of(0L))) > 0) {
-			this.state = this.state.withGenericQC(qc);
+			safetyStateBuilder.qc(qc);
 		}
 
-		// commit phase on vertex's grandparent if there is a newer consecutive 2-chain
-		// keep the highest consecutive 2-chain as the locked QC
-		Vertex parent = vertexStore.getVertex(qc.getVertexMetadata().getId());
-		if (parent == null) {
-			throw new IllegalStateException(String.format(
-				"QC %s has no vertex at %s", qc, qc.getVertexMetadata().getId()));
+		if (qc.getParent() != null
+			&& qc.getParent().getView().compareTo(this.state.getLockedView()) > 0
+			&& qc.getParent().getView().next().equals(qc.getView())) {
+
+			safetyStateBuilder.lockedView(qc.getParent().getView());
 		}
-		// do not go beyond genesis
-		if (!parent.isGenesis()) {
-			boolean twoChain = parent.getQC().getView().next().equals(parent.getView());
-			if (twoChain && parent.getQC().getView().compareTo(this.state.getLockedView()) > 0) {
-				this.state = this.state.withLockedView(parent.getQC().getView());
-			}
 
-			// decide phase on vertex's great-grandparent if there is a newer consecutive 3-chain
-			// return committed aid
-			Vertex grandparent = vertexStore.getVertex(parent.getQC().getVertexMetadata().getId());
-			if (grandparent == null) {
-				throw new IllegalStateException(String.format(
-					"QC %s has no vertex at %s", qc, qc.getVertexMetadata().getId()));
-			}
-			// do not go beyond genesis
-			if (!grandparent.isGenesis() && twoChain) {
-				boolean threeChain = qc.getVertexMetadata().getId().equals(parent.getId())
-					&& parent.getQC().getVertexMetadata().getId().equals(grandparent.getId())
-					&& grandparent.getQC().getView().next().equals(grandparent.getView());
-				if (threeChain && grandparent.getQC().getView().compareTo(this.state.getCommittedView()) > 0) {
-					this.state = this.state.withCommittedView(grandparent.getQC().getView());
-					return Optional.of(grandparent.getQC().getVertexMetadata().getId());
-				}
+		final Optional<Hash> commitHash;
```

Comment:
Please document the inner workings of this, and maybe even add a note where this is from for future reference.

Reply:
Might be a bit more elegant and brief as:
```
		final Optional<Hash> commitHash = qc.getCommitted().flatMap(vmd -> {
			if (vmd.getView().compareTo(this.state.getCommittedView()) > 0) {
				safetyStateBuilder.committedView(vmd.getView());
				return Optional.of(vmd.getId());
			}
			return Optional.empty();
		});
```